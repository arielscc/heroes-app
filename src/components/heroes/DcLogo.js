export const DcLogo = () => {
  return (
    <div className="bg-white absolute -top-8 right-3 rounded-full p-1 justify-self-end">
      <svg
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 151.47 151.47"
        className="w-14 h-14"
      >
        <g fillRule="evenodd" clipRule="evenodd" fill="#1f2937">
          <path d="M75.6 0C33.83.08-.09 34.13 0 75.88c.08 41.75 34.13 75.67 75.88 75.6 41.75-.11 75.67-34.14 75.6-75.89C151.38 33.84 117.33-.08 75.58 0zM46 16.99l-2.64 9.39 8.01 5.56-9.63.15-2.72 9.26-3.18-9.26-9.81.11 7.67-5.82-3.15-9 7.75 5.48zm86.3 13.04l-2.64 9.4 8.02 5.55-9.66.16-2.7 9.26-3.17-9.26-9.82.13 7.68-5.85-3.18-9 7.78 5.48zm-56.67 2.54a43.18 43.18 0 11.16 86.36 43.18 43.18 0 01-.16-86.36zm-39.74 70.9l-2.64 9.37 8.02 5.56-9.64.13-2.72 9.26-3.18-9.23-9.79.1 7.68-5.82-3.18-9 7.75 5.49zm83.93 10.46l-2.65 9.4 8.02 5.55-9.63.16-2.73 9.26-3.17-9.26-9.79.1 7.67-5.82-3.17-9 7.75 5.48z" />
          <path d="M55.48 59.76l-17.7 13.5 29.92 39.31 17.33-13.2 1.46-10.64-21.06-27.62zm3.65 12.54l15.19 19.95-4.93 3.76-.63-.8L54.2 76.07zM68.79 50.52l-1.2 8.84 21.65 28.42 10.35 1.4 11.45-8.73 1.46-10.64-7.68-10.05-9.52 7.27 4.42 5.82-4.95 3.81-14.98-19.7 4.93-3.77 4.44 5.83 9.55-7.25-7.25-9.53-9.97-1.37-12.7 9.68z" />
        </g>
      </svg>
    </div>
  );
};
