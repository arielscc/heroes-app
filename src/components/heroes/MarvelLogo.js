export const MarvelLogo = () => {
  return (
    <div className="bg-white absolute -top-8 right-3 rounded-full p-1">
      <svg
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 592.3 569.2"
        className="w-14 h-14 mx-auto"
      >
        <g
          transform="matrix(1.17075 0 0 1.17075 -3.7 -14.8)"
          fillRule="evenodd"
          clipRule="evenodd"
        >
          <ellipse
            cx="256.1"
            cy="255.8"
            rx="239.3"
            ry="229.5"
            stroke="#1f2937"
            strokeWidth="27.3"
          />
          <ellipse cx="256.1" cy="255.8" fill="#fff" rx="192.5" ry="184.6" />
          <ellipse
            cx="256.1"
            cy="255.8"
            rx="135.3"
            ry="129.8"
            stroke="#1f2937"
            strokeWidth="15.4"
          />
          <g stroke="#1f2937" strokeWidth="13.6">
            <path
              fill="#fff"
              d="M279.3 44L101.6 434.4h68.3l46.5-112.6h62.1l.8 38.8 59-2.3-4.7-313.5z"
            />
            <path
              fill="none"
              d="M280 262v-88.4L242 262zM280.1 236.4l57.4 58.3M279.3 347.4l58.2-45.8"
            />
          </g>
        </g>
      </svg>
    </div>
  );
};
