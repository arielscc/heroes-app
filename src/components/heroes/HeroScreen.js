import React, { useMemo } from 'react';
import { Redirect, useHistory, useParams } from 'react-router-dom';
import { getHeroById } from '../../selectors/getHeroById';
import { loadImage } from '../helpers/heroImages';
import { DcLogo } from './DcLogo';
import { MarvelLogo } from './MarvelLogo';

export const HeroScreen = () => {
  const { heroeId } = useParams();
  // const hero = getHeroById(heroeId);

  const history = useHistory();

  const hero = useMemo(() => getHeroById(heroeId), [heroeId]);
  if (!hero) {
    return <Redirect to="/" />;
  }

  const { superhero, publisher, alter_ego, first_appearance, characters } =
    hero;

  return (
    <div className="md:flex md:w-2/3 m-auto rounded-2xl overflow-hidden bg-hero">
      <div className="md:w-3/6 xl:w-2/6">
        <img
          src={loadImage(`${heroeId}.jpg`)}
          alt=""
          className="w-full h-full"
        />
      </div>
      <div className="text-center p-4 bg-hero text-white relative md:w-3/6 md:p-14 xl:w-4/6 xl:p-20 ">
        {publisher === 'DC Comics' ? (
          <div className="md:relative md:-mr-4">
            <DcLogo />
          </div>
        ) : (
          <div className="md:relative md:-mr-4">
            <MarvelLogo />
          </div>
        )}
        <h1 className="font-bold text-4xl mt-4 leading-8 md:text-6xl">
          {superhero}
        </h1>
        <p className="uppercase text-gray-400 text-sm tracking-wide">
          {publisher}
        </p>
        <div className="text-left mt-2 font-light flex flex-col">
          <div className="flex justify-between my-1">
            <p className="font-bold">Name:</p>
            <p> {alter_ego}</p>
          </div>
          <hr className="border-gray-600" />
          <div className="flex justify-between my-1">
            <p className="font-bold">First Apparance:</p>
            <p className="text-right w-1/2"> {first_appearance}</p>
          </div>
          <hr className="border-gray-600" />
          <div className="flex justify-between my-1">
            <p className="font-bold">Characters:</p>
            <p> {characters}</p>
          </div>
          <hr className="border-gray-600" />
          <div className="flex justify-between my-1">
            <p>
              Lorem ipsum dolor sit amet consectetur, adipisicing elit. Minima
              porro impedit placeat accusamus, beatae dolor maiores reiciendis
              voluptates, obcaecati inventore optio doloremque provident ex
              expedita dolorem aliquam doloribus nemo dignissimos? Lorem ipsum
              dolor sit amet consectetur adipisicing elit. Amet ad molestiae
              placeat? Facilis nemo quidem libero repellendus fuga officia,
              repudiandae dignissimos deleniti, recusandae aperiam iure, ducimus
              possimus ipsum atque rerum?
            </p>
          </div>
          <button
            onClick={() => {
              if (history.length <= 2) {
                console.log('push');
                return history.push('/');
              } else {
                console.log('goback');
                return history.goBack();
              }
              // return history.length <= 2 ? history.push('/') : history.goBack();
            }}
            className="self-end py-2 px-4 m-5 rounded bg-blue-600 hover:bg-blue-800 font-bold"
          >
            Go back
          </button>
        </div>
      </div>
    </div>
  );
};
