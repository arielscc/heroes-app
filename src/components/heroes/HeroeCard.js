import React from 'react';
import { Link } from 'react-router-dom';
import { DcLogo } from './DcLogo';
import { MarvelLogo } from './MarvelLogo';

import { loadImage } from '../helpers/heroImages';

export const HeroeCard = ({ hero }) => {
  const { id, superhero, publisher, alter_ego, first_appearance, characters } =
    hero;

  return (
    <Link
      to={`hero/${id}`}
      className="relative rounded-xl transform hover:scale-105 cursor-pointer duration-150 overflow-hidden image"
    >
      <img
        src={loadImage(`${id}.jpg`)}
        alt={superhero}
        className="rounded-xl"
      />
      <div className="absolute bg-hero rounded-b-xl left-0 right-0 bottom-0  p-2 border-t-4 border-white flex footer duration-150">
        {publisher === 'DC Comics' ? <DcLogo /> : <MarvelLogo />}
        <div className="self-auto text-gray-200 ml-2 mb-2 title">
          <h2 className="text-xl font-extrabold text-gray-100">{superhero}</h2>
          <p className="uppercase text-xs mb-4 text-gray-400">{alter_ego}</p>
          <p>{publisher}</p>
          <p>{first_appearance}</p>
          <p className="font-thin text-sm text-gray-400">{characters}</p>
        </div>
      </div>
    </Link>
  );
};
