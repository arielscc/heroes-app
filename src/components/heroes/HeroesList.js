import React, { useMemo } from 'react';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroeCard } from './HeroeCard';

export const HeroesList = ({ publisher }) => {
  // const heroes = getHeroesByPublisher(publisher);

  const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher]);

  return (
    <div className="w-10/12 mx-auto grid 2xl:grid-cols-5 xl:grid-cols-4 lg:grid-cols-3 md:grid-cols-3 sm:grid-cols-2 gap-6">
      {heroes.map((hero) => {
        return <HeroeCard key={hero.id} hero={hero} />;
      })}
    </div>
  );
};
