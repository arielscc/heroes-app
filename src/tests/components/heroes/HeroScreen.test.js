// import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import { MemoryRouter, Route, Router, useHistory } from 'react-router';
import { HeroScreen } from '../../../components/heroes/HeroScreen';

describe('Probando el componente <HeroScreen/>', () => {
  test('debe de mostrar el componente redirect si no hay argumentos en el URL', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <HeroScreen />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Redirect').exists()).toBe(true);
  });

  test('debe de mostrar un hero si el parametro existe y se encuentra', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-hulk']}>
        <Route path="/hero/:heroeId" component={HeroScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.contains('Hulk')).toBe(true);
  });
  test('debe de regresar a la pantalla anteior con push', () => {
    // TODO: -Esta prueba queda inconclusa debido a que no se encuentra forma de pasar los metodos del useHistory al componente <HeroScreen/>

    const mockHistoryPush = jest.fn();
    const mockHistoryGoBack = jest.fn();

    jest.mock('react-router-dom', () => ({
      ...jest.requireActual('react-router-dom'),
      useHistory: () => ({
        length: 1,
        push: mockHistoryPush,
        goBack: mockHistoryGoBack,
      }),
    }));

    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-hulk']}>
        <Route path="/hero/:heroeId" component={HeroScreen} />
      </MemoryRouter>
    );
    // wrapper.find('button').prop('onClick')();
    // expect(mockHistoryGoBack).toHaveBeenCalled();
  });
});
