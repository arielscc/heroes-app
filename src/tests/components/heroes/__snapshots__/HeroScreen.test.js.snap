// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente <HeroScreen/> debe de mostrar el componente redirect si no hay argumentos en el URL 1`] = `""`;

exports[`Probando el componente <HeroScreen/> debe de mostrar un hero si el parametro existe y se encuentra 1`] = `
<div
  className="md:flex md:w-2/3 m-auto rounded-2xl overflow-hidden bg-hero"
>
  <div
    className="md:w-3/6 xl:w-2/6"
  >
    <img
      alt=""
      className="w-full h-full"
      src=""
    />
  </div>
  <div
    className="text-center p-4 bg-hero text-white relative md:w-3/6 md:p-14 xl:w-4/6 xl:p-20 "
  >
    <div
      className="md:relative md:-mr-4"
    >
      <div
        className="bg-white absolute -top-8 right-3 rounded-full p-1"
      >
        <svg
          className="w-14 h-14 mx-auto"
          viewBox="0 0 592.3 569.2"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clipRule="evenodd"
            fillRule="evenodd"
            transform="matrix(1.17075 0 0 1.17075 -3.7 -14.8)"
          >
            <ellipse
              cx="256.1"
              cy="255.8"
              rx="239.3"
              ry="229.5"
              stroke="#1f2937"
              strokeWidth="27.3"
            />
            <ellipse
              cx="256.1"
              cy="255.8"
              fill="#fff"
              rx="192.5"
              ry="184.6"
            />
            <ellipse
              cx="256.1"
              cy="255.8"
              rx="135.3"
              ry="129.8"
              stroke="#1f2937"
              strokeWidth="15.4"
            />
            <g
              stroke="#1f2937"
              strokeWidth="13.6"
            >
              <path
                d="M279.3 44L101.6 434.4h68.3l46.5-112.6h62.1l.8 38.8 59-2.3-4.7-313.5z"
                fill="#fff"
              />
              <path
                d="M280 262v-88.4L242 262zM280.1 236.4l57.4 58.3M279.3 347.4l58.2-45.8"
                fill="none"
              />
            </g>
          </g>
        </svg>
      </div>
    </div>
    <h1
      className="font-bold text-4xl mt-4 leading-8 md:text-6xl"
    >
      Hulk
    </h1>
    <p
      className="uppercase text-gray-400 text-sm tracking-wide"
    >
      Marvel Comics
    </p>
    <div
      className="text-left mt-2 font-light flex flex-col"
    >
      <div
        className="flex justify-between my-1"
      >
        <p
          className="font-bold"
        >
          Name:
        </p>
        <p>
           
          Bruce Banner
        </p>
      </div>
      <hr
        className="border-gray-600"
      />
      <div
        className="flex justify-between my-1"
      >
        <p
          className="font-bold"
        >
          First Apparance:
        </p>
        <p
          className="text-right w-1/2"
        >
           
          The Incredible Hulk #1
        </p>
      </div>
      <hr
        className="border-gray-600"
      />
      <div
        className="flex justify-between my-1"
      >
        <p
          className="font-bold"
        >
          Characters:
        </p>
        <p>
           
          Bruce Banner
        </p>
      </div>
      <hr
        className="border-gray-600"
      />
      <div
        className="flex justify-between my-1"
      >
        <p>
          Lorem ipsum dolor sit amet consectetur, adipisicing elit. Minima porro impedit placeat accusamus, beatae dolor maiores reiciendis voluptates, obcaecati inventore optio doloremque provident ex expedita dolorem aliquam doloribus nemo dignissimos? Lorem ipsum dolor sit amet consectetur adipisicing elit. Amet ad molestiae placeat? Facilis nemo quidem libero repellendus fuga officia, repudiandae dignissimos deleniti, recusandae aperiam iure, ducimus possimus ipsum atque rerum?
        </p>
      </div>
      <button
        className="self-end py-2 px-4 m-5 rounded bg-blue-600 hover:bg-blue-800 font-bold"
        onClick={[Function]}
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;
